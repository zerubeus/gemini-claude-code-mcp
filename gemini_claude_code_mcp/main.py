"""Main entry point for Gemini-Claude Code MCP server."""

import sys
from typing import Optional

import click
from rich.console import Console

from gemini_claude_code_mcp.mcp_server.server import mcp
from gemini_claude_code_mcp.utils.logging import get_logger, setup_logging

console = Console()


@click.command()
@click.option(
    '--log-level',
    type=click.Choice(['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']),
    default='INFO',
    help='Set logging level',
)
@click.option('--log-file', type=click.Path(), default=None, help='Log to file (in addition to console)')
@click.option('--no-rich', is_flag=True, default=False, help='Disable rich console output')
def main(log_level: str, log_file: Optional[str], no_rich: bool):
    """Run the Gemini-Claude Code MCP server."""
    # Set up logging
    setup_logging(level=log_level, log_file=log_file, use_rich=not no_rich)

    logger = get_logger(__name__)

    # Welcome message
    if not no_rich:
        console.print('[bold cyan]ðŸš€ Gemini-Claude Code MCP Server[/bold cyan]', justify='center')
        console.print("Bridging Claude Code with Google Gemini's massive context window", justify='center', style='dim')
        console.print()

    logger.info(
        'Starting Gemini-Claude Code MCP server', log_level=log_level, log_file=log_file, rich_output=not no_rich
    )

    try:
        # Run the FastMCP server
        mcp.run()
    except KeyboardInterrupt:
        logger.info('Server stopped by user')
        if not no_rich:
            console.print('\n[yellow]Server stopped by user[/yellow]')
    except Exception as e:
        logger.error('Server error', error=str(e), exc_info=True)
        if not no_rich:
            console.print(f'\n[red]Server error: {e}[/red]')
        sys.exit(1)


if __name__ == '__main__':
    main()
