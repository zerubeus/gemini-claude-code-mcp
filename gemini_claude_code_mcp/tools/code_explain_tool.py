"""AMP Envelope tools for controlling AMP EG parameters on the Moog Sub 37."""

from typing import Any

from fastmcp import FastMCP


def register_code_explain_tools(mcp: FastMCP[Any]):
    """Register all tools related to code explanation and documentation generation."""

    @mcp.tool()
    def generate_doc_for_specific_folder(absolute_folder_path: str, prompt: str = 'Generate documentation'):  # type: ignore
        """Generate documentation for a specific folder.

        Args:
            absolute_folder_path (str): The absolute folder path to generate documentation for.
            prompt (str): The prompt to use. Defaults to "Generate documentation".

        """
        pass
