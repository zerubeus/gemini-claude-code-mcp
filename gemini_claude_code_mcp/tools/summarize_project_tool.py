"""Project summarization tool for analyzing codebases."""

from pathlib import Path
from typing import Any

from fastmcp import FastMCP

from gemini_claude_code_mcp.models.context import AnalysisRequest, ChunkingStrategy, FilePattern
from gemini_claude_code_mcp.services.file_collector import FileCollector
from gemini_claude_code_mcp.services.large_context_analyzer import LargeContextAnalyzer
from gemini_claude_code_mcp.utils.chunking import count_tokens
from gemini_claude_code_mcp.utils.logging import get_logger

logger = get_logger(__name__)


def register_summarize_project_tool(mcp: FastMCP[Any]):
    """Register the summarize_project tool with the MCP server."""

    @mcp.tool()
    async def summarize_project(  # type: ignore
        directory_path: str,
        focus_areas: list[str] | None,
        include_patterns: list[str] | None,
        exclude_patterns: list[str] | None,
    ) -> dict[str, Any]:
        """Generate a comprehensive summary of a project/codebase.

        This tool analyzes an entire project directory and provides a multi-level summary including:
        - Project structure and organization
        - Technology stack and dependencies
        - Main components and their purposes
        - Architecture patterns and design decisions
        - Key APIs and interfaces

        Args:
            directory_path: Absolute path to the project directory
            focus_areas: Optional list of specific areas to focus on (e.g., ["authentication", "api", "database"])
            include_patterns: Optional file patterns to include (e.g., ["*.py", "src/**/*.js"])
            exclude_patterns: Optional file patterns to exclude beyond defaults

        Returns:
            Dictionary containing:
            - overview: High-level project summary
            - structure: Directory structure analysis
            - tech_stack: Detected technologies and frameworks
            - components: Main components and their descriptions
            - architecture: Architectural patterns and decisions
            - statistics: Project statistics (file counts, sizes, etc.)
        """
        try:
            # Validate directory path
            project_path = Path(directory_path).resolve()
            if not project_path.exists():
                return {
                    'error': f'Directory not found: {directory_path}',
                    'status': 'failed',
                }

            if not project_path.is_dir():
                return {
                    'error': f'Path is not a directory: {directory_path}',
                    'status': 'failed',
                }

            logger.info(f'Starting project summary for: {project_path}')

            # Initialize services
            file_collector = FileCollector()
            analyzer = LargeContextAnalyzer()

            # Prepare file pattern
            file_pattern = FilePattern(
                include=include_patterns or [],
                exclude=exclude_patterns or [],
                respect_gitignore=True,
            )

            # Collect files
            logger.info('Collecting project files...')
            collected_files = await file_collector.collect_files(str(project_path), file_pattern)

            if not collected_files:
                return {
                    'error': 'No files found matching the specified patterns',
                    'status': 'failed',
                }

            logger.info(f'Collected {len(collected_files)} files')

            # Generate project structure
            structure = _generate_project_structure(project_path, collected_files)

            # Prepare analysis query
            focus_context = ''
            if focus_areas:
                focus_context = f'\nPay special attention to these areas: {", ".join(focus_areas)}'

            analysis_query = f"""Analyze this codebase and provide a comprehensive project summary including:
            1. **Overview**: A high-level description of what this project does and its main purpose
            2. **Technology Stack**: List all detected programming languages, frameworks, libraries, and tools
            3. **Architecture**: Identify architectural patterns, design decisions, and project organization
            4. **Main Components**: List and describe the main components, modules, or services
            5. **Key Features**: Identify the key features and functionalities implemented
            6. **Dependencies**: Note important external dependencies and integrations
            7. **Code Quality**: Comment on code organization, patterns, and best practices observed
            {focus_context}
            Provide a structured, detailed summary that would help a developer quickly understand this codebase."""

            # Combine file contents for analysis
            combined_content = _combine_file_contents(collected_files, project_path)
            total_tokens = count_tokens(combined_content)

            logger.info(f'Total content tokens: {total_tokens}')

            # Create analysis request
            analysis_request = AnalysisRequest(
                query=analysis_query,
                content=combined_content,
                chunking_strategy=ChunkingStrategy.CODE_AWARE,
                context_metadata={
                    'project_path': str(project_path),
                    'file_count': len(collected_files),
                },
            )

            # Analyze with LargeContextAnalyzer
            logger.info('Analyzing project content...')
            analysis_result = await analyzer.analyze(analysis_request)

            # Generate statistics
            statistics = _generate_statistics(collected_files)

            # Parse the analysis response into structured sections
            structured_summary = _parse_analysis_response(analysis_result.response or 'No analysis available')

            return {
                'status': 'success',
                'project_path': str(project_path),
                'overview': structured_summary.get('overview', 'No overview available'),
                'structure': structure,
                'tech_stack': structured_summary.get('tech_stack', {}),
                'architecture': structured_summary.get('architecture', {}),
                'components': structured_summary.get('components', []),
                'key_features': structured_summary.get('key_features', []),
                'dependencies': structured_summary.get('dependencies', []),
                'code_quality': structured_summary.get('code_quality', {}),
                'statistics': statistics,
                'analysis_details': {
                    'files_analyzed': len(collected_files),
                    'total_tokens': total_tokens,
                    'used_gemini': analysis_result.used_gemini,
                    'chunks_processed': analysis_result.chunks_processed,
                },
            }

        except Exception as e:
            logger.error(f'Error summarizing project: {e}', exc_info=True)
            return {
                'error': f'Failed to summarize project: {str(e)}',
                'status': 'failed',
            }


def _generate_project_structure(project_path: Path, collected_files: list[Any]) -> dict[str, Any]:
    """Generate a hierarchical representation of the project structure."""
    structure: dict[str, Any] = {'name': project_path.name, 'type': 'directory', 'children': {}}

    for file in collected_files:
        relative_path = Path(file.relative_path)
        parts = relative_path.parts

        current: dict[str, Any] = structure['children']
        for part in parts[:-1]:
            if part not in current:
                current[part] = {'type': 'directory', 'children': {}}
            current = current[part]['children']

        # Add the file
        filename = parts[-1]
        current[filename] = {
            'type': 'file',
            'language': file.language,
            'size': file.size,
            'tokens': file.token_count,
        }

    return structure


def _combine_file_contents(collected_files: list[Any], project_path: Path) -> str:
    """Combine file contents with metadata headers."""
    combined_parts: list[str] = []

    # Add project overview
    combined_parts.append(f'# Project: {project_path.name}')
    combined_parts.append(f'# Path: {project_path}')
    combined_parts.append(f'# Total Files: {len(collected_files)}')
    combined_parts.append('\n---\n')

    # Add each file with header
    for file in collected_files:
        header = f'\n### File: {file.relative_path}\n'
        header += f'Language: {file.language or "unknown"}\n'
        header += f'Size: {file.size} bytes | Tokens: {file.token_count}\n'
        header += '```\n'

        combined_parts.append(header)
        combined_parts.append(file.content)
        combined_parts.append('\n```\n')

    return '\n'.join(combined_parts)


def _generate_statistics(collected_files: list[Any]) -> dict[str, Any]:
    """Generate project statistics."""
    stats: dict[str, Any] = {
        'total_files': len(collected_files),
        'total_size_bytes': sum(f.size for f in collected_files),
        'total_tokens': sum(f.token_count for f in collected_files),
        'languages': {},
        'file_types': {},
    }

    # Count by language
    for file in collected_files:
        lang = file.language or 'unknown'
        if lang not in stats['languages']:
            stats['languages'][lang] = 0
        stats['languages'][lang] += 1

        # Count by extension
        ext = Path(file.relative_path).suffix or 'no_extension'
        if ext not in stats['file_types']:
            stats['file_types'][ext] = 0
        stats['file_types'][ext] += 1

    # Sort languages by count
    stats['languages'] = dict(sorted(stats['languages'].items(), key=lambda x: x[1], reverse=True))

    # Add human-readable size
    size_mb = stats['total_size_bytes'] / (1024 * 1024)
    stats['total_size_mb'] = round(size_mb, 2)

    return stats


def _parse_analysis_response(response: str) -> dict[str, Any]:
    """Parse the analysis response into structured sections."""
    # This is a simple parser that looks for markdown sections
    # In production, you might want to use a more sophisticated approach

    sections: dict[str, Any] = {
        'overview': '',
        'tech_stack': {},
        'architecture': {},
        'components': [],
        'key_features': [],
        'dependencies': [],
        'code_quality': {},
    }

    current_section: str | None = None
    current_content: list[str] = []

    lines = response.split('\n')

    for line in lines:
        # Check for section headers
        if line.startswith('**Overview**') or line.startswith('1. **Overview**'):
            current_section = 'overview'
            current_content = []
        elif line.startswith('**Technology Stack**') or line.startswith('2. **Technology Stack**'):
            current_section = 'tech_stack'
            current_content = []
        elif line.startswith('**Architecture**') or line.startswith('3. **Architecture**'):
            current_section = 'architecture'
            current_content = []
        elif line.startswith('**Main Components**') or line.startswith('4. **Main Components**'):
            current_section = 'components'
            current_content = []
        elif line.startswith('**Key Features**') or line.startswith('5. **Key Features**'):
            current_section = 'key_features'
            current_content = []
        elif line.startswith('**Dependencies**') or line.startswith('6. **Dependencies**'):
            current_section = 'dependencies'
            current_content = []
        elif line.startswith('**Code Quality**') or line.startswith('7. **Code Quality**'):
            current_section = 'code_quality'
            current_content = []
        elif current_section:
            # Add content to current section
            if line.strip():
                current_content.append(line.strip())

    # Process collected content
    if 'overview' in sections:
        sections['overview'] = '\n'.join(current_content).strip()

    # For now, return the full response in overview if parsing fails
    if not any(sections.values()):
        sections['overview'] = response

    return sections
