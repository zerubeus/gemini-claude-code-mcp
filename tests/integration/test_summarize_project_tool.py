"""Integration test for summarize_project tool."""

import json
from pathlib import Path
from typing import Any

import pytest
from fastmcp import Client, FastMCP
from syrupy.assertion import SnapshotAssertion

from gemini_claude_code_mcp.tools.summarize_project_tool import register_summarize_project_tool


@pytest.fixture
def mcp_server() -> FastMCP[Any]:
    """Create MCP server with summarize_project tool registered."""
    server: FastMCP[Any] = FastMCP('TestServer')
    register_summarize_project_tool(server)
    return server


@pytest.fixture
def test_project_dir(tmp_path: Path) -> Path:
    """Create a minimal test project structure."""
    # Create project structure
    project_dir = tmp_path / 'test_project'
    project_dir.mkdir()

    # Create some Python files
    src_dir = project_dir / 'src'
    src_dir.mkdir()

    # Main module
    main_py_content = """\"\"\"Main module for test project.\"\"\"

    from typing import List


    class Calculator:
        \"\"\"Simple calculator class for testing.\"\"\"

        def add(self, a: int, b: int) -> int:
            \"\"\"Add two numbers.\"\"\"
            return a + b

        def multiply(self, a: int, b: int) -> int:
            \"\"\"Multiply two numbers.\"\"\"
            return a * b


    def process_items(items: List[str]) -> List[str]:
        \"\"\"Process a list of items.\"\"\"
        return [item.upper() for item in items]


    if __name__ == "__main__":
        calc = Calculator()
        print(calc.add(5, 3))
    """
    (src_dir / 'main.py').write_text(main_py_content)

    # Utils module
    utils_py_content = """\"\"\"Utility functions for the project.\"\"\"

    import json
    from pathlib import Path


    def load_config(config_path: str) -> dict:
        \"\"\"Load configuration from JSON file.\"\"\"
        with open(config_path, 'r') as f:
            return json.load(f)


    def save_data(data: dict, output_path: str) -> None:
        \"\"\"Save data to JSON file.\"\"\"
        with open(output_path, 'w') as f:
            json.dump(data, f, indent=2)


    def format_path(path: str) -> Path:
        \"\"\"Convert string path to Path object.\"\"\"
        return Path(path).resolve()
    """
    (src_dir / 'utils.py').write_text(utils_py_content)

    # Create tests directory
    tests_dir = project_dir / 'tests'
    tests_dir.mkdir()

    test_main_content = """\"\"\"Tests for main module.\"\"\"

    from src.main import Calculator, process_items


    def test_calculator_add():
        calc = Calculator()
        assert calc.add(2, 3) == 5


    def test_process_items():
        items = ['hello', 'world']
        result = process_items(items)
        assert result == ['HELLO', 'WORLD']
    """
    (tests_dir / 'test_main.py').write_text(test_main_content)

    # Create configuration files
    pyproject_content = """[project]
    name = "test-project"
    version = "0.1.0"
    description = "A test project for integration testing"
    dependencies = [
        "pytest>=7.0.0",
    ]

    [build-system]
    requires = ["setuptools>=61.0"]
    build-backend = "setuptools.build_meta"
    """
    (project_dir / 'pyproject.toml').write_text(pyproject_content)

    readme_content = """# Test Project

    This is a minimal test project for integration testing the summarize_project tool.

    ## Features

    - Simple calculator implementation
    - Utility functions for file operations
    - Basic test suite

    ## Usage

    ```python
    from src.main import Calculator

    calc = Calculator()
    result = calc.add(10, 20)
    ```
    """
    (project_dir / 'README.md').write_text(readme_content)

    # Add .gitignore
    gitignore_content = """__pycache__/
    *.pyc
    .pytest_cache/
    """
    (project_dir / '.gitignore').write_text(gitignore_content)

    return project_dir


@pytest.mark.asyncio
async def test_summarize_project_basic(
    mcp_server: FastMCP[Any], test_project_dir: Path, snapshot: SnapshotAssertion
) -> None:
    """Test basic project summarization with real Gemini API."""
    async with Client(mcp_server) as client:
        result = await client.call_tool(
            'summarize_project',
            {
                'directory_path': str(test_project_dir),
            },
        )

        # Extract the response
        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        response = json.loads(result[0].text)  # type: ignore  # type: ignore

        # Verify structure
        assert response['status'] == 'success'
        assert response['project_path'] == str(test_project_dir)
        assert 'overview' in response
        assert 'structure' in response
        assert 'statistics' in response
        assert 'analysis_details' in response

        # Verify statistics are accurate
        stats = response['statistics']
        assert stats['total_files'] == 3  # main.py, utils.py, test_main.py (only .py files by default)
        assert stats['languages']['python'] == 3

        # Verify structure
        structure = response['structure']
        assert structure['name'] == 'test_project'
        assert 'src' in structure['children']
        assert 'main.py' in structure['children']['src']['children']
        assert 'utils.py' in structure['children']['src']['children']

        # Create a deterministic snapshot by removing variable data
        snapshot_data = {
            'status': response['status'],
            'structure': response['structure'],
            'statistics': response['statistics'],
            'has_overview': bool(response.get('overview')),
            'has_tech_stack': bool(response.get('tech_stack')),
            'has_architecture': bool(response.get('architecture')),
            'has_components': bool(response.get('components')),
            'analysis_details': {
                'files_analyzed': response['analysis_details']['files_analyzed'],
                'used_gemini': response['analysis_details']['used_gemini'],
            },
        }

        # Compare with snapshot
        assert snapshot_data == snapshot


@pytest.mark.asyncio
async def test_summarize_project_with_focus_areas(
    mcp_server: FastMCP[Any], test_project_dir: Path, snapshot: SnapshotAssertion
) -> None:
    """Test project summarization with focus areas."""
    async with Client(mcp_server) as client:
        result = await client.call_tool(
            'summarize_project',
            {
                'directory_path': str(test_project_dir),
                'focus_areas': ['testing', 'calculator'],
                'include_patterns': ['*.py'],
            },
        )

        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        response = json.loads(result[0].text)  # type: ignore

        assert response['status'] == 'success'
        assert response['statistics']['total_files'] == 3  # Only Python files

        # Create snapshot data
        snapshot_data = {
            'status': response['status'],
            'files_analyzed': response['analysis_details']['files_analyzed'],
            'total_files': response['statistics']['total_files'],
            'languages': response['statistics']['languages'],
        }

        assert snapshot_data == snapshot


@pytest.mark.asyncio
async def test_summarize_project_invalid_directory(mcp_server: FastMCP[Any]) -> None:
    """Test error handling for invalid directory."""
    async with Client(mcp_server) as client:
        result = await client.call_tool(
            'summarize_project',
            {
                'directory_path': '/non/existent/directory',
            },
        )

        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        response = json.loads(result[0].text)  # type: ignore

        assert response['status'] == 'failed'
        assert 'error' in response
        assert 'Directory not found' in response['error']


@pytest.mark.asyncio
async def test_summarize_project_empty_directory(mcp_server: FastMCP[Any], tmp_path: Path) -> None:
    """Test handling of empty directory."""
    empty_dir = tmp_path / 'empty_project'
    empty_dir.mkdir()

    async with Client(mcp_server) as client:
        result = await client.call_tool(
            'summarize_project',
            {
                'directory_path': str(empty_dir),
            },
        )

        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        response = json.loads(result[0].text)  # type: ignore

        assert response['status'] == 'failed'
        assert 'No files found' in response['error']


@pytest.fixture
def large_test_project_dir(tmp_path: Path) -> Path:
    """Create a large test project that exceeds Claude's context limit."""
    project_dir = tmp_path / 'large_test_project'
    project_dir.mkdir()

    # Create many Python files with substantial content
    src_dir = project_dir / 'src'
    src_dir.mkdir()

    # Generate enough content to exceed 200k tokens but stay under 1M
    # Each file will have ~15000 tokens, so we need about 15 files to get ~225k tokens
    large_file_content = """
# This is a large file with substantial content to test Gemini integration
""" + '\n'.join(
        [
            f"""
class Component{i}:
    '''A complex component with many methods and substantial documentation.
    
    This component handles various operations including data processing,
    validation, transformation, and persistence. It implements multiple
    design patterns and integrates with various external services.
    '''
    
    def __init__(self, config: dict):
        '''Initialize the component with configuration.
        
        Args:
            config: Configuration dictionary containing all necessary parameters
                   for initializing this component including database connections,
                   API endpoints, timeout values, and feature flags.
        '''
        self.config = config
        self.data = []
        self.cache = {{}}
        self.connections = []
        self.validators = []
        self.transformers = []
        self.handlers = []
        self.middleware = []
        self.plugins = []
        self.listeners = []
        self.observers = []
    
    def process_data(self, input_data: list) -> list:
        '''Process input data through multiple transformation stages.
        
        This method implements a complex data processing pipeline that includes:
        - Data validation and sanitization
        - Type conversion and normalization
        - Business logic application
        - Error handling and recovery
        - Performance optimization
        - Caching and memoization
        - Logging and monitoring
        
        Args:
            input_data: Raw input data to be processed
            
        Returns:
            Processed data ready for consumption
        '''
        # Extensive processing logic here
        result = []
        for item in input_data:
            # Validate
            if not self._validate(item):
                continue
            
            # Transform
            transformed = self._transform(item)
            
            # Apply business logic
            processed = self._apply_business_logic(transformed)
            
            # Cache results
            self._cache_result(item, processed)
            
            result.append(processed)
        
        return result
    
    def _validate(self, item: Any) -> bool:
        '''Validate an item against all registered validators.'''
        for validator in self.validators:
            if not validator.validate(item):
                return False
        return True
    
    def _transform(self, item: Any) -> Any:
        '''Transform an item through all registered transformers.'''
        result = item
        for transformer in self.transformers:
            result = transformer.transform(result)
        return result
    
    def _apply_business_logic(self, item: Any) -> Any:
        '''Apply complex business logic to the item.'''
        # Simulate complex processing
        result = item
        for handler in self.handlers:
            result = handler.handle(result)
        return result
    
    def _cache_result(self, key: Any, value: Any) -> None:
        '''Cache the processing result for future use.'''
        self.cache[str(key)] = value
    
    # Add many more methods to increase token count...
    {
                ' '.join(
                    [
                        f'''
    def method_{j}(self, param{j}: Any) -> Any:
        """Another method with documentation to increase token count.
        
        This method performs operation {j} on the input parameter and returns
        the processed result after applying various transformations and validations.
        """
        return param{j}
    '''
                        for j in range(50)  # Increased methods per class
                    ]
                )
            }
"""
            for i in range(10)  # Generate more classes per file
        ]
    )

    # Create 15 files to exceed token limit but stay under Gemini's limit
    for i in range(15):
        file_path = src_dir / f'component_{i}.py'
        file_path.write_text(large_file_content)

    return project_dir


@pytest.mark.asyncio
async def test_summarize_large_project_uses_gemini(
    mcp_server: FastMCP[Any], large_test_project_dir: Path, snapshot: SnapshotAssertion
) -> None:
    """Test that large projects trigger Gemini usage."""
    async with Client(mcp_server) as client:
        result = await client.call_tool(
            'summarize_project',
            {
                'directory_path': str(large_test_project_dir),
            },
        )

        assert hasattr(result[0], 'text'), 'Result should have text attribute'
        response = json.loads(result[0].text)  # type: ignore

        assert response == snapshot

        # Verify the request succeeded
        assert response['status'] == 'success'

        # Most importantly: verify Gemini was actually used!
        assert response['analysis_details']['used_gemini'] is True
        assert response['analysis_details']['chunks_processed'] > 0

        # Verify we processed a large number of files
        assert response['statistics']['total_files'] == 15
        assert response['analysis_details']['total_tokens'] > 200000  # Exceeds Claude's limit
        assert response['analysis_details']['total_tokens'] < 1000000  # But stays under Gemini's limit
