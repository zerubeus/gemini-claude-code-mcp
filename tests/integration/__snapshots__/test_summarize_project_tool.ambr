# serializer version: 1
# name: test_summarize_large_project_uses_gemini
  dict({
    'analysis_details': dict({
      'chunks_processed': 1,
      'files_analyzed': 15,
      'total_tokens': 527981,
      'used_gemini': True,
    }),
    'architecture': dict({
    }),
    'code_quality': dict({
    }),
    'components': list([
    ]),
    'dependencies': list([
    ]),
    'key_features': list([
    ]),
    'overview': '''
      Here is a comprehensive analysis of the provided codebase.
      
      ### **Project Summary: large_test_project**
      
      #### 1. Overview
      
      This project, `large_test_project`, is a Python-based system designed for modular, configurable, and extensible data processing. Its core purpose is to provide a framework for creating complex data processing pipelines. Data flows through a series of stages including validation, transformation, and business logic application. The system is architected with a strong emphasis on component-based design, allowing for different functionalities to be plugged in at runtime via configuration. Although the codebase appears to be a large-scale skeleton or a synthetic test project with significant code duplication, it clearly outlines a sophisticated architectural intent.
      
      #### 2. Technology Stack
      
      *   **Programming Language**:
          *   **Python**: The entire codebase is written in Python.
      *   **Frameworks & Libraries**:
          *   No external frameworks (like Django, Flask, etc.) or libraries are detected. The project uses standard Python features.
          *   **`typing`**: The standard library `typing` module is used for type hints (`dict`, `list`, `Any`), which enhances code clarity and allows for static analysis.
      
      #### 3. Architecture
      
      The project's architecture is its most defining characteristic, showcasing several well-known design patterns.
      
      *   **Architectural Patterns**:
          *   **Component-Based Architecture**: The system is fundamentally built on components. The code is organized into `component_*.py` files, each containing multiple `Component` classes. This promotes modularity and separation of concerns.
          *   **Pipeline Pattern**: The `process_data` method in each component implements a classic data processing pipeline. Data is passed sequentially through distinct stages: validation, transformation, business logic, and finally caching.
          *   **Strategy / Plugin Pattern**: The behavior of each component is made extensible by using lists of handlers (`validators`, `transformers`, `handlers`). This allows new functionality (e.g., a new validation rule) to be "plugged in" by adding it to the appropriate list, likely via the configuration object, without modifying the component's core code.
          *   **Observer Pattern**: The presence of `listeners` and `observers` lists strongly suggests an event-driven mechanism where components can subscribe to and react to events within the system.
          *   **Middleware Pattern**: The `middleware` list hints at a design where requests or data can be processed by a chain of independent handlers before reaching the main business logic.
      
      *   **Design Decisions**:
          *   **Configuration-Driven**: Components are initialized with a `config` dictionary, indicating that their behavior, connections, and plugins are injected at runtime rather than being hardcoded. This makes the system highly flexible and adaptable.
          *   **Extensibility**: The plugin-based design is central to the architecture, allowing developers to extend the system's capabilities easily.
          *   **Modularity**: Although the implementation is repetitive, the intent is to have discrete, independent components that handle specific parts of the data processing workflow.
      
      *   **Project Organization**:
          *   The source code is contained within a `src` directory.
          *   The file and class naming convention (`component_0.py`, `Component1`, etc.) is numerical and unconventional, suggesting the project may be auto-generated for testing or demonstration purposes.
      
      #### 4. Main Components
      
      The codebase consists of 15 files, each containing 10 identically structured `Component` classes. Due to this uniformity, we can describe a generic component.
      
      *   **`component_*.py` / `Component*` Class**:
          *   **Description**: This is the fundamental building block of the system. Each component acts as a configurable data processor.
          *   **Initialization (`__init__`)**: It is initialized with a `config` dictionary. It sets up several empty lists to hold plugins for various pipeline stages: `validators`, `transformers`, `handlers`, `middleware`, `plugins`, `listeners`, and `observers`. It also initializes a `cache` dictionary.
          *   **Data Processing (`process_data`)**: This is the main public method. It takes a list of data, iterates through it, and for each item, executes the validation, transformation, and business logic steps by calling its private helper methods. Processed results are cached.
          *   **Helper Methods (`_validate`, `_transform`, `_apply_business_logic`)**: These private methods iterate over their respective plugin lists (e.g., `self.validators`) and apply each plugin to the data item.
      
      #### 5. Key Features
      
      *   **Configurable Data Processing Pipeline**: The core feature is the ability to process data through a multi-stage pipeline.
      *   **Data Validation & Sanitization**: The pipeline includes a dedicated validation stage to ensure data integrity.
      *   **Data Transformation & Normalization**: Data can be transformed or normalized before business logic is applied.
      *   **Pluggable Business Logic**: Core logic is encapsulated in "handlers" that can be dynamically configured.
      *   **Result Caching**: Processed results are cached (`self.cache`) to optimize performance for repeated operations on the same data.
      *   **Extensible Framework**: The architecture is designed to be easily extended with new components, plugins, middleware, and listeners.
      
      #### 6. Dependencies
      
      *   **External Dependencies**: None are detected. The project is self-contained and relies only on the Python standard library.
      *   **Internal Dependencies**: The `typing` module is used for type annotations. There are no explicit import relationships between the component files themselves, suggesting they are designed to be independent modules orchestrated by a higher-level application layer.
      
      #### 7. Code Quality
      
      *   **Positive Aspects**:
          *   **Use of Design Patterns**: The code demonstrates a clear and intentional use of powerful design patterns (Component-Based, Pipeline, Strategy), which is excellent for building a scalable and maintainable system.
          *   **Documentation**: Classes and methods are well-documented with docstrings that explain their purpose, arguments, and return values.
          *   **Type Hinting**: The use of type hints improves code readability and maintainability.
      
      *   **Areas for Improvement**:
          *   **Extreme Code Duplication**: The most significant issue is the massive amount of duplicated code. Every `Component` class in every file is identical. In a real-world scenario, this would be addressed by creating a single base `Component` class and inheriting from it. This repetition strongly indicates that the codebase is synthetic and not intended for production in its current state.
          *   **Unconventional Naming**: The numerical naming scheme for files and classes (`component_1.py`, `Component2`) is not descriptive and would make a real project very difficult to navigate and maintain.
          *   **Skeleton Implementation**: The core logic is a skeleton. The `process_data` method outlines a pipeline, but the actual `validators`, `transformers`, and `handlers` are never instantiated or defined, and the numerous `method_*` functions are empty placeholders.
    ''',
    'project_path': '/private/var/folders/7l/y36rc8k13ds07ntsc0q4xnyr0000gn/T/pytest-of-zerbasta/pytest-17/test_summarize_large_project_u0/large_test_project',
    'statistics': dict({
      'file_types': dict({
        '.py': 15,
      }),
      'languages': dict({
        'python': 15,
      }),
      'total_files': 15,
      'total_size_bytes': 2936175,
      'total_size_mb': 2.8,
      'total_tokens': 527460,
    }),
    'status': 'success',
    'structure': dict({
      'children': dict({
        'src': dict({
          'children': dict({
            'component_0.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_1.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_10.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_11.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_12.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_13.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_14.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_2.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_3.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_4.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_5.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_6.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_7.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_8.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
            'component_9.py': dict({
              'language': 'python',
              'size': 195745,
              'tokens': 35164,
              'type': 'file',
            }),
          }),
          'type': 'directory',
        }),
      }),
      'name': 'large_test_project',
      'type': 'directory',
    }),
    'tech_stack': dict({
    }),
  })
# ---
# name: test_summarize_project_basic
  dict({
    'analysis_details': dict({
      'files_analyzed': 3,
      'used_gemini': False,
    }),
    'has_architecture': False,
    'has_components': False,
    'has_overview': True,
    'has_tech_stack': False,
    'statistics': dict({
      'file_types': dict({
        '.py': 3,
      }),
      'languages': dict({
        'python': 3,
      }),
      'total_files': 3,
      'total_size_bytes': 1503,
      'total_size_mb': 0.0,
      'total_tokens': 345,
    }),
    'status': 'success',
    'structure': dict({
      'children': dict({
        'src': dict({
          'children': dict({
            'main.py': dict({
              'language': 'python',
              'size': 602,
              'tokens': 138,
              'type': 'file',
            }),
            'utils.py': dict({
              'language': 'python',
              'size': 571,
              'tokens': 130,
              'type': 'file',
            }),
          }),
          'type': 'directory',
        }),
        'tests': dict({
          'children': dict({
            'test_main.py': dict({
              'language': 'python',
              'size': 330,
              'tokens': 77,
              'type': 'file',
            }),
          }),
          'type': 'directory',
        }),
      }),
      'name': 'test_project',
      'type': 'directory',
    }),
  })
# ---
# name: test_summarize_project_with_focus_areas
  dict({
    'files_analyzed': 3,
    'languages': dict({
      'python': 3,
    }),
    'status': 'success',
    'total_files': 3,
  })
# ---
